### DISCUSSION TITLE
cffi - local installation of new packages

### DISCUSSION THREAD

**Discussion by balrog-nona** [OpenIndiana/oi-userland#6566] [Q&A] [answered]
Hello,

I'm experiencing very strange behaviour when I try to install via `pfexec pkg install -v` newly build cffi packages.
The component after update and adding Python 3.7 and 3.9 looks like this:
https://github.com/OpenIndiana/oi-userland/pull/6503/commits/c9390a4b3681e6c781ba3ef3a5c9a38bbc0d0ae5

Ii builds, installs and publishes fine. Tests look okay too - only 14 failed out of 2000.

But when I try to install it locally, I get this message:
```
pkg install: Package 'library/python/pyopenssl-35' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/library/python/pyopenssl-35@16.2.0-2020.0.1.4
  Reason:  No version matching 'require' dependency library/python/cryptography-35 can be installed
Package 'developer/opensolaris/osnet' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/developer/opensolaris/osnet@0.5.11-2020.0.1.20365
  Reason:  No version matching 'require' dependency developer/build/onbld can be installed
    ----------------------------------------
    Reject:  pkg://openindiana.org/developer/build/onbld@0.5.11-2020.0.1.20365
    Reason:  No version matching 'require' dependency package/pkg can be installed
      ----------------------------------------
      Reject:  pkg://openindiana.org/package/pkg@0.5.11-2020.0.1.5549
      Reason:  No version matching 'require' dependency library/python/cryptography-35 can be installed
      ----------------------------------------
    ----------------------------------------
Package 'package/pkg/zones-proxy' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/package/pkg/zones-proxy@0.5.11-2020.0.1.5549
  Reason:  No version matching 'require' dependency package/pkg/system-repository@0.5.11-2020.0.1.5543 can be installed
    ----------------------------------------
    Reject:  pkg://openindiana.org/package/pkg/system-repository@0.5.11-2020.0.1.5549
    Reason:  No version matching 'require' dependency package/pkg@0.5.11-2020.0.1.5543 can be installed
    ----------------------------------------
Package 'library/python/cryptography-27' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/library/python/cryptography-27@2.6.1-2020.0.1.3
  Reason:  No version matching 'incorporate' dependency library/python/cffi-27@1.12.2 can be installed
Package 'system/zones/brand/ipkg' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/system/zones/brand/ipkg@0.5.11-2020.0.1.5549
  Reason:  No version matching 'require' dependency package/pkg/zones-proxy can be installed
  Reason:  No version matching 'require' dependency package/pkg@0.5.11-2020.0.1.5543 can be installed
Package 'library/python/cryptography' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/library/python/cryptography@2.6.1-2020.0.1.3
  Reason:  No version matching 'conditional' dependency library/python/cryptography-27@2.6.1-2020.0.1.3 can be installed
Package 'package/pkg' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/package/pkg@0.5.11-2020.0.1.5549
  Reason:  [already rejected; see above]
Package 'library/python/pyopenssl' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/library/python/pyopenssl@16.2.0-2020.0.1.4
  Reason:  No version matching 'conditional' dependency library/python/pyopenssl-27@16.2.0-2020.0.1.4 can be installed
    ----------------------------------------
    Reject:  pkg://openindiana.org/library/python/pyopenssl-27@16.2.0-2020.0.1.4
    Reason:  No version matching 'require' dependency library/python/cryptography-27 can be installed
    ----------------------------------------
Package 'library/python/cryptography-35' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/library/python/cryptography-35@2.6.1-2020.0.1.3
  Reason:  No version matching 'incorporate' dependency library/python/cffi-35@1.12.2 can be installed
Package 'developer/build/onbld' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/developer/build/onbld@0.5.11-2020.0.1.20365
  Reason:  [already rejected; see above]
Package 'metapackages/build-essential' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/metapackages/build-essential@1.0-2020.0.1.22
  Reason:  No version matching 'require' dependency developer/build/onbld can be installed
Package 'package/pkg/system-repository' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/package/pkg/system-repository@0.5.11-2020.0.1.5549
  Reason:  [already rejected; see above]
Package 'library/python/pyopenssl-27' must be uninstalled or upgraded if the requested operation is to be performed.
  Reject:  pkg://openindiana.org/library/python/pyopenssl-27@16.2.0-2020.0.1.4
  Reason:  [already rejected; see above]
```
First, I though I have messy build environment, so I changed my BE to the very original after installing OI, updating, setting build environment again and creating BE for installing new components. It didn't work.

Then, I uninstalled packages entire and userland-incorporation. It didn't work either. The message above it the one I got after uninstalling those packages.

The message confuses me because cffi is cryptography's dependency, not the other way around. And the same with pkg - it needs cryptography as a dependency, not the other way around.

Can you tell from the message what is wrong? I have both publishers - openindiana.org and userlad set as non-sticky.
Thank you!

Edit: I discovered that I get the same message when I try to install it on my other OI machine. Therefore, I believe there must be something wrong with the cffi itself.

**Comment by AndWac** [comment, marked as answer]
Am 10.03.21 um 14:35 schrieb Nona Hansel:
>
> Hello,
>
> I'm experiencing very strange behaviour when I try to install via
> |pfexec pkg install -v| newly build cffi packages.
> The component after update and adding Python 3.7 and 3.9 looks like this:
> c9390a4
> <https://github.com/OpenIndiana/oi-userland/commit/c9390a4b3681e6c781ba3ef3a5c9a38bbc0d0ae5>
>
> Ii builds, installs and publishes fine. Tests look okay too - only 14
> failed out of 2000.
>
> But when I try to install it locally, I get this message:
>
> |pkg install: Package 'library/python/pyopenssl-35' must be
> uninstalled or upgraded if the requested operation is to be performed.
> Reject:
> ***@***.***
> Reason: No version matching 'require' dependency
> library/python/cryptography-35 can be installed Package
> 'developer/opensolaris/osnet' must be uninstalled or upgraded if the
> requested operation is to be performed. Reject:
> ***@***.***
> Reason: No version matching 'require' dependency developer/build/onbld
> can be installed ---------------------------------------- Reject:
> ***@***.***
> Reason: No version matching 'require' dependency package/pkg can be
> installed ---------------------------------------- Reject:
> ***@***.*** Reason: No
> version matching 'require' dependency library/python/cryptography-35
> can be installed ----------------------------------------
> ---------------------------------------- Package
> 'package/pkg/zones-proxy' must be uninstalled or upgraded if the
> requested operation is to be performed. Reject:
> ***@***.***
> Reason: No version matching 'require' dependency
> ***@***.*** can be installed
> ---------------------------------------- Reject:
> ***@***.***
> Reason: No version matching 'require' dependency
> ***@***.*** can be installed
> ---------------------------------------- Package
> 'library/python/cryptography-27' must be uninstalled or upgraded if
> the requested operation is to be performed. Reject:
> ***@***.***
> Reason: No version matching 'incorporate' dependency
> ***@***.*** can be installed Package
> 'system/zones/brand/ipkg' must be uninstalled or upgraded if the
> requested operation is to be performed. Reject:
> ***@***.***
> Reason: No version matching 'require' dependency
> package/pkg/zones-proxy can be installed Reason: No version matching
> 'require' dependency ***@***.*** can be installed
> Package 'library/python/cryptography' must be uninstalled or upgraded
> if the requested operation is to be performed. Reject:
> ***@***.***
> Reason: No version matching 'conditional' dependency
> ***@***.*** can be installed
> Package 'package/pkg' must be uninstalled or upgraded if the requested
> operation is to be performed. Reject:
> ***@***.*** Reason:
> [already rejected; see above] Package 'library/python/pyopenssl' must
> be uninstalled or upgraded if the requested operation is to be
> performed. Reject:
> ***@***.***
> Reason: No version matching 'conditional' dependency
> ***@***.*** can be installed
> ---------------------------------------- Reject:
> ***@***.***
> Reason: No version matching 'require' dependency
> library/python/cryptography-27 can be installed
> ---------------------------------------- Package
> 'library/python/cryptography-35' must be uninstalled or upgraded if
> the requested operation is to be performed. Reject:
> ***@***.***
> Reason: No version matching 'incorporate' dependency
> ***@***.*** can be installed Package
> 'developer/build/onbld' must be uninstalled or upgraded if the
> requested operation is to be performed. Reject:
> ***@***.***
> Reason: [already rejected; see above] Package
> 'metapackages/build-essential' must be uninstalled or upgraded if the
> requested operation is to be performed. Reject:
> ***@***.***
> Reason: No version matching 'require' dependency developer/build/onbld
> can be installed Package 'package/pkg/system-repository' must be
> uninstalled or upgraded if the requested operation is to be performed.
> Reject:
> ***@***.***
> Reason: [already rejected; see above] Package
> 'library/python/pyopenssl-27' must be uninstalled or upgraded if the
> requested operation is to be performed. Reject:
> ***@***.***
> Reason: [already rejected; see above] |
>
> First, I though I have messy build environment, so I changed my BE to
> the very original after installing OI, updating, setting build
> environment again and creating BE for installing new components. It
> didn't work.
>
> Then, I uninstalled packages entire and userland-incorporation. It
> didn't work either. The message above it the one I got after
> uninstalling those packages.
>
> The message confuses me because cffi is cryptography dependency, not
> the other way around. And the same with pkg - it needs cryptography as
> a dependency, not the other way around.
>
> Can you tell from the message what is wrong? I have both publishers -
> openindiana.org and userlad set as non-sticky.
> Thank you!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/discussions/6566>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADET4N6IGYN2WARACV3O26DTC5YRTANCNFSM4Y6A7X7A>.
>
Hi,

I think I found the source of your problem:
The package "cryptography" has a "required" dependency on cffi with an
explicit version in cryptography-PYVER.p5m:
...
# Dependencies on cffi must be incorporated.
depend type=incorporate ***@***.***
...


You have created a new version of cffi but missed the explicit
dependency. So you have to update the cryptography package, too.

Regards,
Andreas

**Reply by balrog-nona** [reply]
Thank you, Andreas, now I see it in cryptography-PYVER.p5m

**Comment by balrog-nona** [comment]
Hello again,

I'm still facing some problems with cryptography's update. When I add Python 3.7 and 3.9 into cryptography's Makefile and when I edit cryptography-PYVER.p5m into
```
#Dependencies on cffi must be incorporated.
depend type=incorporate fmri=library/python/cffi-$(PYV)@1.14.0
```
I'll get this message:
```
(cd /export/home/nona/oi-userland/components/python/cryptography/cryptography-2.6.1 ; /usr/bin/env HOME=/export/home/nona/oi-userland/components/python/cryptography/build/config-amd64-3.7 LD_OPTIONS="-M /usr/lib/ld/map.noexstk -M /usr/lib/ld/map.noexdata -M /usr/lib/ld/map.pagealign -Bdirect -z ignore" LD_EXEC_OPTIONS="-z aslr=disable"  CC="/usr/gcc/7/bin/gcc" CFLAGS="-m64 -O3 " PKG_CONFIG_PATH="/usr/openssl/1.0/lib/64/pkgconfig:/usr/lib/amd64/pkgconfig" \
        /usr/bin/python3.7 ./setup.py build )
WARNING: The wheel package is not available.
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3.7 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-wheel-pr7prms7/cffi_5b87a93dbdc745a69464afcd22b42f7b/setup.py'"'"'; __file__='"'"'/tmp/pip-wheel-pr7prms7/cffi_5b87a93dbdc745a69464afcd22b42f7b/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-5l4mdtd0
       cwd: /tmp/pip-wheel-pr7prms7/cffi_5b87a93dbdc745a69464afcd22b42f7b/
  Complete output (12 lines):
  ld: fatal: file _configtest.o: wrong ELF class: ELFCLASS64
  ld: fatal: file processing errors. No output written to _configtest
  collect2: error: ld returned 1 exit status
  Note: will not use '__sync_synchronize()' in the C code
  ***** The above error message can be safely ignored.
  
  usage: setup.py [global_opts] cmd1 [cmd1_opts] [cmd2 [cmd2_opts] ...]
     or: setup.py --help [cmd1 cmd2 ...]
     or: setup.py --help-commands
     or: setup.py cmd --help
  
  error: invalid command 'bdist_wheel'
  ----------------------------------------
  ERROR: Failed building wheel for cffi
ERROR: Failed to build one or more wheels
Traceback (most recent call last):
  File "/usr/lib/python3.7/vendor-packages/setuptools/installer.py", line 126, in fetch_build_egg
    subprocess.check_call(cmd)
  File "/usr/lib/python3.7/subprocess.py", line 363, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/usr/bin/python3.7', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmp1k5iccyu', '--quiet', 'cffi!=1.11.3,>=1.8']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./setup.py", line 324, in <module>
    **keywords_with_side_effects(sys.argv)
  File "/usr/lib/python3.7/vendor-packages/setuptools/__init__.py", line 152, in setup
    _install_setup_requires(attrs)
  File "/usr/lib/python3.7/vendor-packages/setuptools/__init__.py", line 147, in _install_setup_requires
    dist.fetch_build_eggs(dist.setup_requires)
  File "/usr/lib/python3.7/vendor-packages/setuptools/dist.py", line 676, in fetch_build_eggs
    replace_conflicting=True,
  File "/usr/lib/python3.7/vendor-packages/pkg_resources/__init__.py", line 766, in resolve
    replace_conflicting=replace_conflicting
  File "/usr/lib/python3.7/vendor-packages/pkg_resources/__init__.py", line 1049, in best_match
    return self.obtain(req, installer)
  File "/usr/lib/python3.7/vendor-packages/pkg_resources/__init__.py", line 1061, in obtain
    return installer(requirement)
  File "/usr/lib/python3.7/vendor-packages/setuptools/dist.py", line 732, in fetch_build_egg
    return fetch_build_egg(self, req)
  File "/usr/lib/python3.7/vendor-packages/setuptools/installer.py", line 128, in fetch_build_egg
    raise DistutilsError(str(e)) from e
distutils.errors.DistutilsError: Command '['/usr/bin/python3.7', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/tmp/tmp1k5iccyu', '--quiet', 'cffi!=1.11.3,>=1.8']' returned non-zero exit status 1.
gmake: *** [/export/home/nona/oi-userland/make-rules/setup.py.mk:93: /export/home/nona/oi-userland/components/python/cryptography/build/amd64-3.7/.built] Error 1
```

I belive that the problem is that the cffi hasn't been installed locally. The easiest way to get rid of this issue would be not to update cffi package but it wasn't possible for me to just add python 3.7 and 3.9 into existing version because it gave me an error for Python 3.9 which I couldn't fix.

**Comment by AndWac** [comment]
Am 30.03.21 um 18:01 schrieb Nona Hansel:
>
> Hello again,
>
> I'm still facing some problems with cryptography's update. When I add
> Python 3.7 and 3.9 into cryptography's Makefile and when I edit
> cryptography-PYVER.p5m into
>
> |#Dependencies on cffi must be incorporated. depend type=incorporate
> ***@***.*** |
>
> I'll get this message:
>
> |(cd
> /export/home/nona/oi-userland/components/python/cryptography/cryptography-2.6.1
> ; /usr/bin/env
> HOME=/export/home/nona/oi-userland/components/python/cryptography/build/config-amd64-3.7
> LD_OPTIONS="-M /usr/lib/ld/map.noexstk -M /usr/lib/ld/map.noexdata -M
> /usr/lib/ld/map.pagealign -Bdirect -z ignore" LD_EXEC_OPTIONS="-z
> aslr=disable" CC="/usr/gcc/7/bin/gcc" CFLAGS="-m64 -O3 "
> PKG_CONFIG_PATH="/usr/openssl/1.0/lib/64/pkgconfig:/usr/lib/amd64/pkgconfig"
> \ /usr/bin/python3.7 ./setup.py build ) WARNING: The wheel package is
> not available. ERROR: Command errored out with exit status 1: command:
> /usr/bin/python3.7 -u -c 'import sys, setuptools, tokenize;
> sys.argv[0] =
> '"'"'/tmp/pip-wheel-pr7prms7/cffi_5b87a93dbdc745a69464afcd22b42f7b/setup.py'"'"';
> __file__='"'"'/tmp/pip-wheel-pr7prms7/cffi_5b87a93dbdc745a69464afcd22b42f7b/setup.py'"'"';f=getattr(tokenize,
> '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"',
> '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))'
> bdist_wheel -d /tmp/pip-wheel-5l4mdtd0 cwd:
> /tmp/pip-wheel-pr7prms7/cffi_5b87a93dbdc745a69464afcd22b42f7b/
> Complete output (12 lines): ld: fatal: file _configtest.o: wrong ELF
> class: ELFCLASS64 ld: fatal: file processing errors. No output written
> to _configtest collect2: error: ld returned 1 exit status Note: will
> not use '__sync_synchronize()' in the C code ***** The above error
> message can be safely ignored. usage: setup.py [global_opts] cmd1
> [cmd1_opts] [cmd2 [cmd2_opts] ...] or: setup.py --help [cmd1 cmd2 ...]
> or: setup.py --help-commands or: setup.py cmd --help error: invalid
> command 'bdist_wheel' ---------------------------------------- ERROR:
> Failed building wheel for cffi ERROR: Failed to build one or more
> wheels Traceback (most recent call last): File
> "/usr/lib/python3.7/vendor-packages/setuptools/installer.py", line
> 126, in fetch_build_egg subprocess.check_call(cmd) File
> "/usr/lib/python3.7/subprocess.py", line 363, in check_call raise
> CalledProcessError(retcode, cmd) subprocess.CalledProcessError:
> Command '['/usr/bin/python3.7', '-m', 'pip',
> '--disable-pip-version-check', 'wheel', '--no-deps', '-w',
> '/tmp/tmp1k5iccyu', '--quiet', 'cffi!=1.11.3,>=1.8']' returned
> non-zero exit status 1. The above exception was the direct cause of
> the following exception: Traceback (most recent call last): File
> "./setup.py", line 324, in <module>
> **keywords_with_side_effects(sys.argv) File
> "/usr/lib/python3.7/vendor-packages/setuptools/__init__.py", line 152,
> in setup _install_setup_requires(attrs) File
> "/usr/lib/python3.7/vendor-packages/setuptools/__init__.py", line 147,
> in _install_setup_requires dist.fetch_build_eggs(dist.setup_requires)
> File "/usr/lib/python3.7/vendor-packages/setuptools/dist.py", line
> 676, in fetch_build_eggs replace_conflicting=True, File
> "/usr/lib/python3.7/vendor-packages/pkg_resources/__init__.py", line
> 766, in resolve replace_conflicting=replace_conflicting File
> "/usr/lib/python3.7/vendor-packages/pkg_resources/__init__.py", line
> 1049, in best_match return self.obtain(req, installer) File
> "/usr/lib/python3.7/vendor-packages/pkg_resources/__init__.py", line
> 1061, in obtain return installer(requirement) File
> "/usr/lib/python3.7/vendor-packages/setuptools/dist.py", line 732, in
> fetch_build_egg return fetch_build_egg(self, req) File
> "/usr/lib/python3.7/vendor-packages/setuptools/installer.py", line
> 128, in fetch_build_egg raise DistutilsError(str(e)) from e
> distutils.errors.DistutilsError: Command '['/usr/bin/python3.7', '-m',
> 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w',
> '/tmp/tmp1k5iccyu', '--quiet', 'cffi!=1.11.3,>=1.8']' returned
> non-zero exit status 1. gmake: ***
> [/export/home/nona/oi-userland/make-rules/setup.py.mk:93:
> /export/home/nona/oi-userland/components/python/cryptography/build/amd64-3.7/.built]
> Error 1 |
>
> I belive that the problem is that the cffi hasn't been installed
> locally. The easiest way to get rid of this issue would be not to
> update cffi package but it wasn't possible for me to just add python
> 3.7 and 3.9 into existing version because it gave me an error for
> Python 3.9 which I couldn't fix.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/OpenIndiana/oi-userland/discussions/6566#discussioncomment-548886>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADET4NZWD4DCOGA3BCPTCJTTGHYVPANCNFSM4Y6A7X7A>.
>
You have a chicken-and-egg problem (or cyclic dependency).
I suggest to remove the dependency on the exact version. I don't know
why it has been introduced anyways.
And as the original author didn't bother to give an explanation we
should remove it until we find a bug.
The dependency on cffi is modeled in the Makefile so you can completely
remove it from the manifest.
